### PR TITLE
Resolve problems related with ruptures containing NaNs

### PR DESCRIPTION
**Pull request by mmpagani**
This solves an issue with multi ruptures based on Kite Surfaces. While tests are all green, after a more thorough check of the rupture base class I must say that I am not happy with the current structure since there is not a clear separation between the same methods for different typologies of surfaces. We will need to perform a comprehensive review soon.

### CODE REVIEW

File: openquake/hazardlib/geo/surface/base.py [hunk 1]
```diff
@@ -22,10 +22,28 @@
 """
 import numpy
 import math
+from openquake.hazardlib.geo.mesh import Mesh
 from openquake.hazardlib.geo import geodetic, utils, Point, Line,\
     RectangularMesh
 
 
+def _get_finite_mesh(mesh):
+    idx = numpy.isfinite(mesh.lons.flat)
```
Comment (micheles) [inline openquake/hazardlib/geo/surface/base.py:31]: `idx` is an array of booleans, not of indices. In such I use the convention of calling `ok`:
```python
ok = numpy.isfinite(mesh.lons.flat)
if numpy.all(ok):
￼        return mesh
```

File: openquake/hazardlib/geo/surface/base.py [hunk 2]
```diff
@@ -22,10 +22,28 @@
...
+    if numpy.all(idx):
+        return mesh
+    idx = numpy.reshape(idx, mesh.lons.shape)
+    return Mesh(mesh.lons[idx], mesh.lats[idx], mesh.depths[idx])
+
+
+def _get_finite_top_rupture(mesh):
+    idx = numpy.isfinite(mesh.lons.flat)
```
Comment (micheles) [inline openquake/hazardlib/geo/surface/base.py:39]: Same as above

File: openquake/hazardlib/geo/surface/kite_fault.py [hunk 3]
```diff
@@ -161,6 +162,10 @@ def get_dip(self) -> float:
                                   np.zeros_like(self.mesh.depths[1:, col_idx]))
                 vdists = (self.mesh.depths[1:, col_idx] -
                           self.mesh.depths[:-1, col_idx])
+
+                idx = np.logical_and(np.isfinite(hdists), np.isfinite(vdists))
```
Comment (micheles) [inline openquake/hazardlib/geo/surface/kite_fault.py:166]: Same as above

File: openquake/hazardlib/geo/surface/kite_fault.py [hunk 4]
```diff
@@ -182,6 +187,17 @@ def get_strike(self) -> float:
...
 
+    def get_top_edge_depth(self):
+        """
+        Return minimum depth of surface's top edge.
+
+        :returns:
+            Float value, the vertical distance between the earth surface
+            and the shallowest point in surface's top edge in km.
+        """
+        idx = np.isfinite(self.mesh.lons[0, :])
```
Comment (micheles) [inline openquake/hazardlib/geo/surface/kite_fault.py:198]: Same as above

File: openquake/hazardlib/geo/surface/kite_fault.py [hunk 5]
```diff
@@ -349,12 +365,62 @@ def surface_projection(self):
...
+        top_length = np.sqrt(np.sum(top * top, axis=-1))
+        left_length = np.sqrt(np.sum(left * left, axis=-1))
+
+        bottom = along_azimuth[1:]
+        right = updip[:, 1:]
+        br_area = geo_utils.triangle_area(bottom, right, diag)
+        bottom_length = np.sqrt(np.sum(bottom * bottom, axis=-1))
+        right_length = np.sqrt(np.sum(right * right, axis=-1))
+
+        # Remove cells without a fiinite area
```
Comment (micheles) [inline openquake/hazardlib/geo/surface/kite_fault.py:396]: Mispelled `fiinite`

File: openquake/hazardlib/geo/surface/multi.py [hunk 6]
```diff
@@ -277,9 +277,11 @@ def get_top_edge_depth(self):
         average value (in km).
         """
         areas = self._get_areas()
-        depths = numpy.array(
-            [surf.get_top_edge_depth() for surf in self.surfaces])
-        return numpy.sum(areas * depths) / numpy.sum(areas)
+        depths = numpy.array([numpy.mean(surf.get_top_edge_depth()) for surf
+                              in self.surfaces])
+        ted = numpy.sum(areas * depths) / numpy.sum(areas)
+        assert numpy.isfinite(ted)
```
Comment (micheles) [inline openquake/hazardlib/geo/surface/multi.py:283]: This should read
```
assert numpy.isfinite(ted).all()
```

File: openquake/hazardlib/geo/surface/multi.py [hunk 7]
```diff
@@ -309,9 +309,10 @@ def get_dip(self):
         formula for weighted mean is used.
         """
         areas = self._get_areas()
-        dips = numpy.array([surf.get_dip() for surf in self.surfaces])
-
-        return numpy.sum(areas * dips) / numpy.sum(areas)
+        dips = numpy.array([numpy.mean(surf.get_dip()) for surf in self.surfaces])
+        dip = numpy.sum(areas * dips) / numpy.sum(areas)
+        assert numpy.isfinite(dip)
```
Comment (micheles) [inline openquake/hazardlib/geo/surface/multi.py:314]:
``` 
assert numpy.isfinite(dip).all()
```

File: openquake/hazardlib/geo/surface/kite_fault.py [hunk 8]
```diff
@@ -349,12 +365,62 @@ def surface_projection(self):
 
         return plos, plas
```
Comment (micheles) [inline openquake/hazardlib/geo/surface/kite_fault.py:367]: Please add a comment that this method is duplicated

**Comment by mmpagani** [comment]
Thanks for the comments. All addressed in this additional commit.

**Review by micheles — APPROVED** [review]

**Pull request merged**